### PR TITLE
PUBDEV-8612: Speed-up job polling (in R)

### PR DESCRIPTION
**Pull request by michalkurka**
This script show that we can save ~800ms per model training just on polling in R.

```
prostate.hex <- h2o.importFile("smalldata/logreg/prostate.csv")

prostate.hex$CAPSULE <- as.factor(prostate.hex$CAPSULE)
prostate.hex$RACE <- as.factor(prostate.hex$RACE)

ntrees <- 100

for (test in c(FALSE, TRUE)) {
  if (test) {
    h2o.rapids('(setproperty "fast_polling" "true")')
    print("\n FAST POLLING \n")
  } else {
    print("\n REGULAR POLLING \n")
    h2o.rapids('(setproperty "fast_polling" "false")')
  }
  
  for (i in seq(5)) {
    start_time <- Sys.time()
    prostate.h2o <- h2o.gbm(x = 3:9, y = "CAPSULE", training_frame = prostate.hex, distribution = "bernoulli", ntrees = ntrees, max_depth = 5, min_rows = 10, learn_rate = 0.1)
    end_time <- Sys.time()
    
    print("Waited:")
    print(end_time - start_time)
    print("Run-time (ms):")
    print(prostate.h2o@model$run_time)
    print("Wasted time (ms):")
    print(as.numeric(end_time - start_time)*1000 - prostate.h2o@model$run_time)
  }
}
```

Results:

```
[1] "\n REGULAR POLLING \n"
  |======================================================================================================================================================================| 100%
[1] "Waited:"
Time difference of 1.171726 secs
[1] "Run-time (ms):"
[1] 273
[1] "Wasted time (ms):"
[1] 898.726
  |======================================================================================================================================================================| 100%
[1] "Waited:"
Time difference of 1.132804 secs
[1] "Run-time (ms):"
[1] 271
[1] "Wasted time (ms):"
[1] 861.8042
  |======================================================================================================================================================================| 100%
[1] "Waited:"
Time difference of 1.12984 secs
[1] "Run-time (ms):"
[1] 270
[1] "Wasted time (ms):"
[1] 859.8399
  |======================================================================================================================================================================| 100%
[1] "Waited:"
Time difference of 1.129631 secs
[1] "Run-time (ms):"
[1] 272
[1] "Wasted time (ms):"
[1] 857.631
  |======================================================================================================================================================================| 100%
[1] "Waited:"
Time difference of 1.129786 secs
[1] "Run-time (ms):"
[1] 275
[1] "Wasted time (ms):"
[1] 854.786
```

And fast polling:

```
[1] "\n FAST POLLING \n"
  |======================================================================================================================================================================| 100%
[1] "Waited:"
Time difference of 0.374512 secs
[1] "Run-time (ms):"
[1] 279
[1] "Wasted time (ms):"
[1] 95.51196
  |======================================================================================================================================================================| 100%
[1] "Waited:"
Time difference of 0.3655801 secs
[1] "Run-time (ms):"
[1] 271
[1] "Wasted time (ms):"
[1] 94.58008
  |======================================================================================================================================================================| 100%
[1] "Waited:"
Time difference of 0.3723218 secs
[1] "Run-time (ms):"
[1] 280
[1] "Wasted time (ms):"
[1] 92.32184
  |======================================================================================================================================================================| 100%
[1] "Waited:"
Time difference of 0.372376 secs
[1] "Run-time (ms):"
[1] 289
[1] "Wasted time (ms):"
[1] 83.37597
  |======================================================================================================================================================================| 100%
[1] "Waited:"
Time difference of 0.374588 secs
[1] "Run-time (ms):"
[1] 280
[1] "Wasted time (ms):"
[1] 94.58801
```

### CODE REVIEW

**Comment by michalkurka** [comment]
@tomasfryda @sebhrusen WDYT about this approach?

I introduced a concept of Session Properties - each R session can set properties to the corresponding backend Session object. This way I can control some behavior (in this case timed polling) in R while leaving Flow/Python unmodified.

Python has its own complex polling strategy and even though this approach can speed up polling even for Python, it makes the progress bar less smooth (the speed-up is also not as big).

I didn't want to change the polling API because of backward compatibility concerns with this particular API and I think the concept of session properties is useful for the future as well.

Your thoughts?

**Comment by tomasfryda** [comment]
I like it. Only a small note - assuming my experimental PR to use `curl` in R will remain/won't be much modified [this part concerning the headers](https://github.com/h2oai/h2o-3/pull/6081/files#diff-cb411d06ba64487adc9b50df35d1bbb6d33f6ee29a657167dfa4f3765980fb1bR149-R152) should be duplicated to the new [`.h2o.doRawREST.with.curl`](https://github.com/h2oai/h2o-3/blob/master/h2o-r/h2o-package/R/communication.R#L108).

**Comment by michalkurka** [comment]
@tomasfryda got it, thank you - I will make it compatible with your changes

File: h2o-core/src/main/java/water/Job.java [hunk 1]
```diff
@@ -480,4 +481,42 @@ private void update_from_remote( ) {
       }
   }
   @Override public Class<KeyV3.JobKeyV3> makeSchema() { return KeyV3.JobKeyV3.class; }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  protected void copyOver(Job src) {
+    super.copyOver(src);
+    _barrier = src._barrier;
+  }
```
Comment (sebhrusen) [inline h2o-core/src/main/java/water/Job.java:490]: is it not used by the icing/weaving/serialization logic? what would it mean to serialize a transient attribute like a Barrier? Did you try this in multinode?

File: h2o-core/src/main/java/water/Job.java [hunk 2]
```diff
@@ -480,4 +481,42 @@ private void update_from_remote( ) {
...
+      throw new IllegalArgumentException("Must be a Job not a " + ice.getClass());
+    }
+    final Job<?> j = (Job<?>) ice;
+    if (timeMillis > 0 && !j.isStopped()) {
+      j.waitForDone(timeMillis);
+    }
+    return j;
+  }
+
+  void waitForDone(long timeMillis) {
```
Comment (sebhrusen) [inline h2o-core/src/main/java/water/Job.java:508]: I like this idea, maybe the AutoML polling logic could also benefit from it (although I think it should be more radically rethought).

**Review by sebhrusen — COMMENTED** [review]
nice stuff, I like the approach

File: h2o-core/src/main/java/water/Job.java (hunk 1 above)
Comment (sebhrusen) [reply h2o-core/src/main/java/water/Job.java:490]: although if you explicitly overrode this method, there must be a reason… which I don't get immediately

Comment (sebhrusen) [reply h2o-core/src/main/java/water/Job.java:490]: gotcha, in the `tryGetDoneJob` below, you need to retrieve the job by key... and then you need the barrier.
But does it make sense to keep the barrier transient then?

Comment (michalkurka) [reply h2o-core/src/main/java/water/Job.java:490]: the barrier is not h2o-serializable, it is imho not conceptually serializable either - it needs to be transient

File: h2o-core/src/main/java/water/Job.java (hunk 2 above)
Comment (michalkurka) [reply h2o-core/src/main/java/water/Job.java:518]: that is a good point - to my knowledge, AutoML is using an F/J thread for busy waiting

Comment (michalkurka) [reply h2o-core/src/main/java/water/Job.java:518]: quick attempt: https://github.com/h2oai/h2o-3/pull/6093

File: h2o-core/src/main/java/water/Job.java (hunk 1 above)
Comment (sebhrusen) [reply h2o-core/src/main/java/water/Job.java:490]: I agree that conceptually, it's weird to consider the barrier as serializable, but I guess I misunderstand the semantic of `copyOver`:
from Weaver.java:
```java
    // DTasks need to be able to copy all their (non transient) fields from one
    // DTask instance over another, to match the MRTask API.
    if( iced_cc.subclassOf(_dtask) ) {
      String cpbody_impl =
        make_body(icer_cc, iced_cc, iced_clazz, "copyOver", null, null,
                  "  protected void copyOver(water.Freezable fdst, water.Freezable fsrc) {\n",
```
I understood from the comments above that `copyOver` was supposed to copy only non-transient fields, but the few time it's being overridden, it's to copy transient ones, so I'm definitely missing something. Hard to find out exactly how it is used...

Comment (michalkurka) [reply h2o-core/src/main/java/water/Job.java:490]: I need to check it is really necessary - it might not be, thanks for the comment

Comment (michalkurka) [reply h2o-core/src/main/java/water/Job.java:490]: yes, I think that is a misususe (= doesn't do anything) Job is not descendant from DTask so it doesn't apply here

**Comment by michalkurka** [comment]
@tomasfryda can you please review this PR?

**Review by sebhrusen — APPROVED** [review]
Looks great, thanks @michalkurka

**Review by tomasfryda — APPROVED** [review]
Looks good to me, thank you @michalkurka !

**Pull request merged**